client/src/components/Search/Search.js

```diff
@@ -0,0 +1,361 @@
+   *
+   * @param {Object} props
+   */
+  setOverrides(props) {
+    if (props && (
+        !hasFilters(props.filters) || this.props.formSchemaUrl !== props.formSchemaUrl
+      )) {
+      // clear any overrides that may be in place
+      const schemaUrl = (props && props.formSchemaUrl) || this.props.formSchemaUrl;
+      if (schemaUrl) {
+        this.props.actions.schema.setSchemaStateOverrides(schemaUrl, null);
+      }
+    }
+
+    if (props && (hasFilters(props.filters) && props.formSchemaUrl)) {
+      const filters = props.filters || {};
+      const overrides = {
+        fields: Object
+          .keys(filters)
+          .map((name) => {
+            const value = filters[name];
+            return { name, value };
+          }),
+      };
+
+      // set overrides into redux store, so that it can be accessed by FormBuilder with the same
+      // schemaUrl.
+      this.props.actions.schema.setSchemaStateOverrides(props.formSchemaUrl, overrides);
+    }
+  }
+
+  /**
+   * Update the state search term form an input field change event.
+   * @param event onChangeEvent from a input field.
+   */
+  handleChange(event) {
+      this.setState({ searchText: event.target.value });
+  }
+
+  /**
+   * Try to find the input field and focus on it.
+   */
+  focusInput() {
+    if (this.state.display === DISPLAY.NONE) {
+      return;
+    }
+
+    const node = ReactDOM.findDOMNode(this);
+    if (!node) {
+      return;
+    }
+
+    const input = node.querySelector('.search-box__content-field');
+    // check that it doesn't already have focus
+    if (input !== document.activeElement) {
+      input.focus();
+      if (input.select) {
+        input.select();
+      }
+    }
+  }
+
+  /**
+   * Try to find the first form field in the advanced form and focus on it.
+   */
+  focusFirstFormField() {
+    if (this.state.display !== DISPLAY.EXPANDED) {
+      return;
+    }
+
+    const node = ReactDOM.findDOMNode(this);
+    if (!node) {
+      return;
+    }
+
+    const form = node.querySelector('.search-form form');
+    if (!form) {
+      return;
+    }
+
+    const input = form.querySelector('input, textarea, select');
+    if (input) {
+      input.focus();
+      if (input.select) {
+        input.select();
+      }
+    }
+  }
+
+  /**
+   * Clear the search term and any advanced search form data.
+   * @param props
+   */
+  clearFormData(props) {
+    this.setState({ searchText: '' });
+
+    const schemaUrl = (props && props.formSchemaUrl) || this.props.formSchemaUrl;
+    if (schemaUrl) {
+      this.props.actions.schema.setSchemaStateOverrides(schemaUrl, null);
+      this.props.actions.reduxForm.initialize(identifier, {}, Object.keys(this.props.formData));
+      this.props.actions.reduxForm.reset(identifier);
+    }
+  }
+
+  /**
+   * Show the Search box and set the focus on it after a slight delay.
+   */
+  open() {
+    this.show();
+    setTimeout(this.focusInput, 50);
```

Comment:
That got copied over from the original search component in asset-admin. I think the timeout is there because there's an animation when the filters form get toggled. So if you try to set the focus right away, the field won't exist yet.

I'll see if there's an easy way to use a callback instead.

Reply:
I had a look and the timeout is indeed there because immediately after the how is called, the form field won't be available yet. So there's nothing to focus on. The previous method is just state update, so it doesn't have a nice elegant callback I can hook into.

There might be a way to use a proper callback to handle the focus, but it's probably going to make the code more complex. So I'll leave it like that.